python/pyspark/sql/connect/client/reattach.py

```diff
@@ -0,0 +1,226 @@
+                                # to loop to get another. Eventually, there will be a non empty
+                                # iterator, because there's always a ResultComplete at the end
+                                # of the stream.
+                                first_loop = False
+                    return has_next
+            return False
+
+    def _release_execute(self, until_response_id: Optional[str]) -> None:
+        """
+        Inform the server to release the execution.
+
+        This will send an asynchronous RPC which will not block this iterator, the iterator can
+        continue to be consumed.
+
+        Release with untilResponseId informs the server that the iterator has been consumed until
+        and including response with that responseId, and these responses can be freed.
+
+        Release with None means that the responses have been completely consumed and informs the
+        server that the completed execution can be completely freed.
+        """
+        from pyspark.sql.connect.client.core import SparkConnectClient
+        from pyspark.sql.connect.client.core import Retrying
+
+        request = self._create_release_execute_request(until_response_id)
+
+        def target():
+            for attempt in Retrying(
+                can_retry=SparkConnectClient.retry_exception, **self._retry_policy
+            ):
+                with attempt:
+                    self._stub.ReleaseExecute(request)
+
+        threading.Thread(target=target).start()
+
+    def _create_reattach_execute_request(self) -> pb2.ReattachExecuteRequest:
+        release = pb2.ReattachExecuteRequest(
+            session_id=self._initial_request.session_id,
+            user_context=self._initial_request.user_context,
+            operation_id=self._initial_request.operation_id,
+        )
+
+        if self._initial_request.client_type:
+            release.client_type = self._initial_request.client_type
+
+        return release
+
+    def _create_release_execute_request(
+        self, until_response_id: Optional[str]
+    ) -> pb2.ReattachExecuteRequest:
+        release = pb2.ReattachExecuteRequest(
+            session_id=self._initial_request.session_id,
+            user_context=self._initial_request.user_context,
+            operation_id=self._initial_request.operation_id,
+        )
+
+        if self._initial_request.client_type:
+            release.client_type = self._initial_request.client_type
+
+        if not until_response_id:
+            release.release_all.CopyFrom(pb2.ReleaseExecuteRequest.ReleaseAll())
+        else:
+            release.release_util.response_id = until_response_id  # type: ignore[attr-defined]
+
+        return release
+
+    def throw(self, type: Any = None, value: Any = None, traceback: Any = None) -> Any:
+        super().throw(type, value, traceback)
+
+    def close(self) -> None:
+        return super().close()
```

Comment:
on close, it should do release_all if it wasn't done.
in scala the Iterator unfortunately doesn't have a close, but if python has it, it should do it.